File: leetcode-04/solutions/cpp/0101.cpp
class Solution {
 public:
  bool isSymmetric(TreeNode* root) { return helper(root, root); }

 private:
  bool helper(TreeNode* p, TreeNode* q) {
    if (!p || !q) return p == q;

    return p->val == q->val &&
           helper(p->left, q->right) &&
           helper(p->right, q->left);
  }
};

File: leetcode-04/solutions/python3/0101.py
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.helper(root, root)

    def helper(self, p: TreeNode, q: TreeNode) -> bool:
        if not p or not q:
            return p == q

        return p.val == q.val and \
            self.helper(p.left, q.right) and \
            self.helper(p.right, q.left)
